File: src/main/java/com/freya02/botcommands/application/ApplicationCommandListener.java
package com.freya02.botcommands.application;

import com.freya02.botcommands.CooldownScope;
import com.freya02.botcommands.application.context.message.MessageCommandInfo;
import com.freya02.botcommands.application.context.user.UserCommandInfo;
import com.freya02.botcommands.application.slash.SlashCommandInfo;
import com.freya02.botcommands.internal.BContextImpl;
import com.freya02.botcommands.internal.Logging;
import com.freya02.botcommands.internal.RunnableEx;
import com.freya02.botcommands.internal.Usability;
import com.freya02.botcommands.internal.Usability.UnusableReason;
import com.freya02.botcommands.internal.utils.Utils;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.GuildChannel;
import net.dv8tion.jda.api.events.interaction.commands.GenericCommandEvent;
import net.dv8tion.jda.api.events.interaction.commands.MessageContextCommandEvent;
import net.dv8tion.jda.api.events.interaction.commands.SlashCommandEvent;
import net.dv8tion.jda.api.events.interaction.commands.UserContextCommandEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import net.dv8tion.jda.api.interactions.Interaction;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;

import java.util.EnumSet;
import java.util.StringJoiner;
import java.util.concurrent.ExecutorService;

public final class ApplicationCommandListener extends ListenerAdapter {
	private static final Logger LOGGER = Logging.getLogger();
	private final BContextImpl context;

	private int commandThreadNumber = 0;
	private final ExecutorService commandService = Utils.createCommandPool(r -> {
		final Thread thread = new Thread(r);
		thread.setDaemon(false);
		thread.setUncaughtExceptionHandler((t, e) -> Utils.printExceptionString("An unexpected exception happened in a context command thread '" + t.getName() + "':", e));
		thread.setName("Context command thread #" + commandThreadNumber++);

		return thread;
	});

	public ApplicationCommandListener(BContextImpl context) {
		this.context = context;
	}

	@Override
	public void onUserContextCommand(@NotNull UserContextCommandEvent event) {
		if (LOGGER.isTraceEnabled()) {
			LOGGER.trace("Received user command: {}", event.getName());
		}

		runCommand(() -> {
			final UserCommandInfo userCommand = context.findUserCommand(event.getCommandPath());

			if (userCommand == null) {
				event.reply(context.getDefaultMessages(event.getGuild()).getApplicationCommandNotFoundMsg()).queue();
				return;
			}

			if (!canRun(event, userCommand)) return;

			if (userCommand.execute(context, event)) {
				userCommand.applyCooldown(event);
			}
		}, event);
	}

	@Override
	public void onMessageContextCommand(@NotNull MessageContextCommandEvent event) {
		if (LOGGER.isTraceEnabled()) {
			LOGGER.trace("Received message command: {}", event.getName());
		}

		runCommand(() -> {
			final MessageCommandInfo messageCommand = context.findMessageCommand(event.getCommandPath());

			if (messageCommand == null) {
				event.reply(context.getDefaultMessages(event.getGuild()).getApplicationCommandNotFoundMsg()).queue();
				return;
			}

			if (!canRun(event, messageCommand)) return;

			if (messageCommand.execute(context, event)) {
				messageCommand.applyCooldown(event);
			}
		}, event);
	}

	@Override
	public void onSlashCommand(@NotNull SlashCommandEvent event) {
		if (LOGGER.isTraceEnabled()) {
			LOGGER.trace("Received slash command: {}", reconstructCommand(event));
		}

		runCommand(() -> {
			final SlashCommandInfo slashCommand = context.findSlashCommand(CommandPath.of(event.getCommandPath()));

			if (slashCommand == null) {
				event.reply(context.getDefaultMessages(event.getGuild()).getApplicationCommandNotFoundMsg()).queue();
				return;
			}

			if (!canRun(event, slashCommand)) return;

			if (slashCommand.execute(context, event)) {
				slashCommand.applyCooldown(event);
			}
		}, event);
	}

	@NotNull
	public static String reconstructCommand(GenericCommandEvent event) {
		if (event instanceof SlashCommandEvent) {
			return ((SlashCommandEvent) event).getCommandString();
		} else {
			return "/" + event.getName();
		}
	}

	private boolean canRun(@NotNull GenericCommandEvent event, ApplicationCommandInfo applicationCommand) {
		final boolean isNotOwner = !context.isOwner(event.getUser().getIdLong());
		if (event.getGuild() != null) {
			final Usability usability = Usability.of(event, applicationCommand, isNotOwner);

			if (usability.isUnusable()) {
				final var unusableReasons = usability.getUnusableReasons();
				if (unusableReasons.contains(UnusableReason.OWNER_ONLY)) {
					reply(event, this.context.getDefaultMessages(event.getGuild()).getOwnerOnlyErrorMsg());
					
					return false;
				} else if (unusableReasons.contains(UnusableReason.USER_PERMISSIONS)) {
					reply(event, this.context.getDefaultMessages(event.getGuild()).getUserPermErrorMsg());
					
					return false;
				} else if (unusableReasons.contains(UnusableReason.BOT_PERMISSIONS)) {
					final StringJoiner missingBuilder = new StringJoiner(", ");

					//Take needed permissions, extract bot current permissions
					final EnumSet<Permission> missingPerms = applicationCommand.getBotPermissions();
					missingPerms.removeAll(event.getGuild().getSelfMember().getPermissions((GuildChannel) event.getChannel()));

					for (Permission botPermission : missingPerms) {
						missingBuilder.add(botPermission.getName());
					}

					reply(event, String.format(this.context.getDefaultMessages(event.getGuild()).getBotPermErrorMsg(), missingBuilder));
					
					return false;
				}
			}
		}

		if (isNotOwner) {
			final int cooldown = applicationCommand.getCooldown(event, event::getName);
			if (cooldown > 0) {
				if (applicationCommand.getCooldownScope() == CooldownScope.USER) {
					reply(event, String.format(this.context.getDefaultMessages(event.getGuild()).getUserCooldownMsg(), cooldown / 1000.0));
				} else if (applicationCommand.getCooldownScope() == CooldownScope.GUILD) {
					reply(event, String.format(this.context.getDefaultMessages(event.getGuild()).getGuildCooldownMsg(), cooldown / 1000.0));
				} else { //Implicit channel
					reply(event, String.format(this.context.getDefaultMessages(event.getGuild()).getChannelCooldownMsg(), cooldown / 1000.0));
				}

				return false;
			}
		}
		
		return true;
	}

	private void runCommand(RunnableEx code, GenericCommandEvent event) {
		commandService.execute(() -> {
			try {
				code.run();
			} catch (Throwable e) {
				e = Utils.getException(e);

				Utils.printExceptionString("Unhandled exception in thread '" + Thread.currentThread().getName() + "' while executing an application command '" + reconstructCommand(event) + "'", e);
				if (event.isAcknowledged()) {
					event.getHook().sendMessage(context.getDefaultMessages(event.getGuild()).getApplicationCommandErrorMsg()).setEphemeral(true).queue();
				} else {
					event.reply(context.getDefaultMessages(event.getGuild()).getApplicationCommandErrorMsg()).setEphemeral(true).queue();
				}

				context.dispatchException("Exception in application command '" + reconstructCommand(event) + "'", e);
			}
		});
	}

	private void reply(Interaction event, String msg) {
		event.reply(msg)
				.setEphemeral(true)
				.queue(null,
						e -> {
							Utils.printExceptionString("Could not send reply message from application command listener", e);

							context.dispatchException("Could not send reply message from application command listener", e);
						}
				);
	}
}

File: src/main/kotlin/com/freya02/botcommands/internal/commands/application/ApplicationCommandListener.kt
package com.freya02.botcommands.internal.commands.application

import com.freya02.botcommands.api.commands.CommandPath
import com.freya02.botcommands.api.commands.CooldownScope
import com.freya02.botcommands.api.commands.application.ApplicationFilteringData
import com.freya02.botcommands.api.core.annotations.BEventListener
import com.freya02.botcommands.api.core.service.annotations.BService
import com.freya02.botcommands.internal.*
import com.freya02.botcommands.internal.Usability.UnusableReason
import com.freya02.botcommands.internal.core.CooldownService
import dev.minn.jda.ktx.messages.reply_
import kotlinx.coroutines.launch
import mu.KotlinLogging
import net.dv8tion.jda.api.events.interaction.command.GenericCommandInteractionEvent
import net.dv8tion.jda.api.events.interaction.command.MessageContextInteractionEvent
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.events.interaction.command.UserContextInteractionEvent
import java.util.*

@BService
internal class ApplicationCommandListener(private val context: BContextImpl, private val cooldownService: CooldownService) {
    private val logger = KotlinLogging.logger {  }
    private val exceptionHandler = ExceptionHandler(context, logger)

    @BEventListener
    suspend fun onSlashCommand(event: SlashCommandInteractionEvent) {
        logger.trace { "Received slash command: ${reconstructCommand(event)}" }

        context.coroutineScopesConfig.applicationCommandsScope.launch {
            try {
                val slashCommand = CommandPath.of(event.fullCommandName).let {
                    context.applicationCommandsContext.findLiveSlashCommand(event.guild, it)
                        ?: context.applicationCommandsContext.findLiveSlashCommand(null, it)
                        ?: throwUser("A slash command could not be found: ${event.fullCommandName}")
                }

                if (!canRun(event, slashCommand)) return@launch
                slashCommand.execute(event, cooldownService)
            } catch (e: Throwable) {
                handleException(e, event)
            }
        }
    }

    @BEventListener
    suspend fun onUserContextCommand(event: UserContextInteractionEvent) {
        logger.trace { "Received user context command: ${reconstructCommand(event)}" }

        context.coroutineScopesConfig.applicationCommandsScope.launch {
            try {
                val userCommand = event.name.let {
                    context.applicationCommandsContext.findLiveUserCommand(event.guild, it)
                        ?: context.applicationCommandsContext.findLiveUserCommand(null, it)
                        ?: throwUser("A user context command could not be found: ${event.name}")
                }

                if (!canRun(event, userCommand)) return@launch
                userCommand.execute(event, cooldownService)
            } catch (e: Throwable) {
                handleException(e, event)
            }
        }
    }

    @BEventListener
    suspend fun onMessageContextCommand(event: MessageContextInteractionEvent) {
        logger.trace { "Received message context command: ${reconstructCommand(event)}" }

        context.coroutineScopesConfig.applicationCommandsScope.launch {
            try {
                val messageCommand = event.name.let {
                    context.applicationCommandsContext.findLiveMessageCommand(event.guild, it)
                        ?: context.applicationCommandsContext.findLiveMessageCommand(null, it)
                        ?: throwUser("A message context command could not be found: ${event.name}")
                }

                if (!canRun(event, messageCommand)) return@launch
                messageCommand.execute(event, cooldownService)
            } catch (e: Throwable) {
                handleException(e, event)
            }
        }
    }

    private fun handleException(e: Throwable, event: GenericCommandInteractionEvent) {
        exceptionHandler.handleException(event, e, "application command '${reconstructCommand(event)}'")

        val generalErrorMsg = context.getDefaultMessages(event).generalErrorMsg
        when {
            event.isAcknowledged -> event.hook.sendMessage(generalErrorMsg).setEphemeral(true).queue()
            else -> event.reply(generalErrorMsg).setEphemeral(true).queue()
        }
    }

    private fun canRun(event: GenericCommandInteractionEvent, applicationCommand: ApplicationCommandInfo): Boolean {
        val applicationFilteringData = ApplicationFilteringData(context, event, applicationCommand)
        for (applicationFilter in context.applicationConfig.applicationFilters) {
            if (!applicationFilter.isAccepted(applicationFilteringData)) {
                logger.trace("Cancelled application commands due to filter")
                return false
            }
        }

        val isNotOwner = !context.isOwner(event.user.idLong)
        val usability = Usability.of(event, applicationCommand, isNotOwner)
        if (usability.isUnusable) {
            val unusableReasons = usability.unusableReasons
            when {
                UnusableReason.OWNER_ONLY in unusableReasons -> {
                    reply(event, context.getDefaultMessages(event).ownerOnlyErrorMsg)
                    return false
                }
                UnusableReason.NSFW_DISABLED in unusableReasons -> {
                    reply(event, context.getDefaultMessages(event).nsfwDisabledErrorMsg)
                    return false
                }
                UnusableReason.NSFW_ONLY in unusableReasons -> {
                    reply(event, context.getDefaultMessages(event).nsfwOnlyErrorMsg)
                    return false
                }
                UnusableReason.NSFW_DM_DENIED in unusableReasons -> {
                    reply(event, context.getDefaultMessages(event).nsfwdmDeniedErrorMsg)
                    return false
                }
                UnusableReason.USER_PERMISSIONS in unusableReasons -> {
                    reply(event, context.getDefaultMessages(event).userPermErrorMsg)
                    return false
                }
                UnusableReason.BOT_PERMISSIONS in unusableReasons -> {
                    if (event.guild == null) throwInternal("BOT_PERMISSIONS got checked even if guild is null")
                    val missingBuilder = StringJoiner(", ")

                    //Take needed permissions, extract bot current permissions
                    val missingPerms = applicationCommand.botPermissions
                    missingPerms.removeAll(event.guild!!.selfMember.getPermissions(event.guildChannel))
                    for (botPermission in missingPerms) {
                        missingBuilder.add(botPermission.getName())
                    }
                    reply(event, context.getDefaultMessages(event).getBotPermErrorMsg(missingBuilder.toString()))

                    return false
                }
            }
        }

        if (isNotOwner) {
            val cooldown = cooldownService.getCooldown(applicationCommand, event)
            if (cooldown > 0) {
                val messages = context.getDefaultMessages(event)

                when (applicationCommand.cooldownStrategy.scope) {
                    CooldownScope.USER -> reply(event, messages.getUserCooldownMsg(cooldown / 1000.0))
                    CooldownScope.GUILD -> reply(event, messages.getGuildCooldownMsg(cooldown / 1000.0))
                    //Implicit CooldownScope.CHANNEL
                    else -> reply(event, messages.getChannelCooldownMsg(cooldown / 1000.0))
                }

                return false
            }
        }

        return true
    }

    private fun reply(event: GenericCommandInteractionEvent, msg: String) {
        event.reply_(msg, ephemeral = true)
            .queue(null) { exceptionHandler.handleException(event, it, "interaction reply") }
    }

    private fun reconstructCommand(event: GenericCommandInteractionEvent): String {
        return when (event) {
            is SlashCommandInteractionEvent -> event.commandString
            else -> event.name
        }
    }
}